### PR TITLE
Replace raw usize IDs with new types

### PR DESCRIPTION
**Pull request by cptpiepmatz**
# Description

In this PR I replaced most of the raw usize IDs with [newtypes](https://doc.rust-lang.org/rust-by-example/generics/new_types.html). Some other IDs already started using new types and in this PR I did not want to touch them. To make the implementation less repetitive, I made use of a generic `Id<T>` with marker structs. If this lands I would try to move make other IDs also in this pattern.

Also at some places I needed to use `cast`, I'm not sure if the type was incorrect and therefore casting not needed or if actually different ID types intermingle sometimes.

# User-Facing Changes

Probably few, if you got a `DeclId` via a function and placed it later again it will still work. 

# Tests + Formatting

- :green_circle: `toolkit fmt`
- :green_circle: `toolkit clippy`
- :green_circle: `toolkit test`
- :green_circle: `toolkit test stdlib`

# After Submitting

All these internal IDs should become similar.

### CODE REVIEW

**Review by sholderbach — COMMENTED** [review]
Less viral refactor than I thought, thanks for giving this a shot.

Curious if other people have opinions on how the code looks after this change.

File: crates/nu-protocol/src/id.rs [hunk 1]
```diff
@@ -1,12 +1,99 @@
...
+    /// Casts the `Id<T>` into `Id<U>` without changing the inner value.
+    ///
+    /// # Attention
+    /// Ensure the type cast is correct. If the wrong type is used, it may indicate a typing mistake elsewhere.
+    pub const fn cast<U>(self) -> Id<U> {
+        Id {
+            inner: self.inner,
+            _phantom: PhantomData,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (cptpiepmatz) [reply crates/nu-protocol/src/id.rs:41]: We sure can. Which conversions are ok?

File: crates/nu-protocol/src/id.rs [hunk 2]
```diff
@@ -1,12 +1,99 @@
...
+    inner: usize,
+    _phantom: PhantomData<T>,
+}
+
+impl<T> Id<T> {
+    /// Creates a new `Id`.
+    ///
+    /// Using a distinct type like `Id` instead of `usize` helps us avoid mixing plain integers
+    /// with identifiers.
+    pub const fn new(inner: usize) -> Self {
```
[the remark this answers is not in the document]
Comment (cptpiepmatz) [reply crates/nu-protocol/src/id.rs:18]: From what I learned so far is `#[inline]` only a pretty loose hint. But I think we shouldn't have worse runtime performance maybe the build time slightly increases. 

How can I check if that change has performance penalities?

**Comment by fdncred** [comment]
My hot take is that I think it makes it more readable and less readable at the same time. More readable because you know that it's a DeclId, for instance, and less readable because you have to look up DeclId to know that it's a usize.

I could live with it or without it. So, I have a +1 and a -1 vote so I cancel myself out. 🤣

File: crates/nu-protocol/src/id.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (cptpiepmatz) [reply crates/nu-protocol/src/id.rs]: I removed the `cast` method entirely and replaced it with these two:

```rs
// conversion methods, ensure that these conversions are valid

impl BlockId {
    /// Reinterprets a `BlockId` as a `ModuleId`.
    /// 
    /// This conversion is necessary as parsed blocks may be modules.
    #[inline]
    pub const fn into_module_id(self) -> ModuleId {
        ModuleId {
            inner: self.inner,
            _phantom: PhantomData,
        }
    }
}

impl ModuleId {
    /// Reinterprets a `ModuleId` as a `BlockId`.
    /// 
    /// The `Expr::Overlay` requires a `BlockId`, but when we call it, we only have the `ModuleId`.
    pub const fn into_block_id(self) -> BlockId {
        BlockId {
            inner: self.inner,
            _phantom: PhantomData,
        }
    }
}
```

Maybe my doc comment is a bit incorrect. Help me out here.

File: crates/nu-protocol/src/id.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (cptpiepmatz) [reply crates/nu-protocol/src/id.rs:19]: I threw a small benchmark at it and I don't see meaningful differences:

Screenshot 2024-09-27 135806

Screenshot 2024-09-27 140015

[the remark this answers is not in the document]
Comment (sholderbach) [reply crates/nu-protocol/src/id.rs:19]: Yeah should be fine, thanks!

**Comment by cptpiepmatz** [comment]
I used `cast` in three places, one of them was `FmtVar`, it now actually uses `VarId`. Such errors will be fewer in the future with this here.

The other two come from the situation that `Expr::Overlay` uses a `BlockId`. Should it just in general use a `ModuleId`?

File: crates/nu-protocol/src/id.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sholderbach) [reply crates/nu-protocol/src/id.rs]: Sounds plausible to me, @kubouch is the module maestro.

Blocks us from accidental surprises and if necessary we now know what to remove/add in the future.

**Comment by sholderbach** [comment]
> I used `cast` in three places, one of them was `FmtVar`, it now actually uses `VarId`. Such errors will be fewer in the future with this here.

Ok that convinces me that we should go ahead with this change. Good call on your part!



> The other two come from the situation that `Expr::Overlay` uses a `BlockId`. Should it just in general use a `ModuleId`?

Not sure about the context, is this only ever going through the two casts or is this referencing in our block store at some point? Otherwise again probably something that @kubouch can help you the most with, but he is quite busy. Should be fine to land with maybe a todo comment.

**Comment by cptpiepmatz** [comment]
Ah, I just checked the `Expr` enum and it has this variant with this comment:
```rs
Overlay(Option<BlockId>), // block ID of the overlay's origin module
```
So it seems to be the expected id, and that change is from @kubouch.

**Comment by kubouch** [comment]
@cptpiepmatz I think the Expr::Overlay should have ModuleId, not BlockId, it might be a bug. Perhaps changing it and removing these cast methods would be the cleanest way. Casting IDs shouldn't be necessary anywhere.

**Comment by cptpiepmatz** [comment]
> @cptpiepmatz I think the Expr::Overlay should have ModuleId, not BlockId, it might be a bug. Perhaps changing it and removing these cast methods would be the cleanest way. Casting IDs shouldn't be necessary anywhere.

I'll try changing it. 👍🏻

**Comment by cptpiepmatz** [comment]
@sholderbach @kubouch I replaced the `BlockId` inside the `Expr::Overlay` with a `ModuleId`. All tests are green, looks good. We don't have any conversion methods anymore.

**Review by IanManske — COMMENTED** [review]
Thanks for the PR, love the new ID types! Just two notes:
- Can we add a `Display` impl for `Id`? It can just be the id number.
- Should `Id` be `repr(transparent)`?

**Comment by cptpiepmatz** [comment]
> * Can we add a `Display` impl for `Id`? It can just be the id number.

I'm not sure how we should handle this as `RegId` also implements `Display` but via just displaying the number.
(I would like to make `SpanId` and `RegId` in the same style if this gets merged.)

> * Should `Id` be `repr(transparent)`?

I could do that, do we get any advantages from that?

**Comment by IanManske** [comment]
> I could do that, do we get any advantages from that?

Nevermind, seems to be unnecessary after some research.

> I'm not sure how we should handle this as RegId also implements Display but via just displaying the number.
(I would like to make SpanId and RegId in the same style if this gets merged.)

One way would be to add a `FmtReg` struct to `ir/display.rs`. But let's save that for later, if at all.

**Review by IanManske — APPROVED** [review]

**Comment by sholderbach** [comment]
Let's go!

**Pull request merged**